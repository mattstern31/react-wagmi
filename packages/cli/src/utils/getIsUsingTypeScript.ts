import { findUp } from 'find-up'

export async function getIsUsingTypeScript() {
  try {
    const cwd = process.cwd()
    const tsconfig = await findUp('tsconfig.json', { cwd })
    if (tsconfig) return true

    const wagmiConfig = await findUp(['wagmi.config.ts', 'wagmi.config.mts'], {
      cwd,
    })
    if (wagmiConfig) return true

    return false
  } catch {
    return false
  }
}
