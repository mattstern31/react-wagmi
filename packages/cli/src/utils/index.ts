export { findConfig } from './findConfig'
export { format } from './format'
export { getAddressDocString } from './getAddressDocString'
export { getIsUsingTypeScript } from './getIsUsingTypeScript'
export { loadEnv } from './loadEnv'
export {
  getIsPackageInstalled,
  getInstallCommand,
  getPackageManager,
} from './packages'
export { resolveConfig } from './resolveConfig'
