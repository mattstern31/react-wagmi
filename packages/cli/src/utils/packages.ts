import { detect } from 'detect-package-manager'
import { execa } from 'execa'

export async function getIsPackageInstalled({
  packageName,
  cwd = process.cwd(),
}: {
  packageName: string
  cwd?: string
}) {
  try {
    const packageManager = await getPackageManager()
    const command =
      packageManager === 'yarn'
        ? ['list', '--pattern', packageName]
        : ['ls', packageName]
    const { stdout } = await execa(packageManager, command, { cwd })
    if (stdout !== '') return true
    return false
  } catch (error) {
    return false
  }
}

export async function getInstallCommand(packageName: string) {
  const packageManager = await getPackageManager()
  switch (packageManager) {
    case 'yarn':
      return [packageManager, ['add', packageName]] as const
    case 'npm':
      return [packageManager, ['install', '--save', packageName]] as const
    case 'pnpm':
      return [packageManager, ['add', packageName]] as const
  }
}

export async function getPackageManager() {
  const userAgent = process.env.npm_config_user_agent
  if (userAgent) {
    if (userAgent.includes('pnpm')) return 'pnpm'
    // The yarn@^3 user agent includes npm, so yarn must be checked first.
    if (userAgent.includes('yarn')) return 'yarn'
    if (userAgent.includes('npm')) return 'npm'
  }
  return detect()
}
