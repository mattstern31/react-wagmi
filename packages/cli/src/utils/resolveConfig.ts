import { bundleRequire } from 'bundle-require'

import type { Config } from '../config'
import type { MaybeArray } from '../types'

type ResolveConfig = {
  /** Path to config file */
  configPath: string
}

/**
 * Bundles and returns wagmi config object from path.
 */
export async function resolveConfig({
  configPath,
}: ResolveConfig): Promise<MaybeArray<Config>> {
  const res = await bundleRequire({
    filepath: configPath,
    ...(process.env.NODE_ENV === 'test'
      ? {
          // FIXME: Workaround https://github.com/egoist/bundle-require/issues/6
          require: (id: string) => import(id),
        }
      : {}),
  })
  let config = res.mod.default
  if (config.default) config = config.default
  if (typeof config !== 'function') return config
  return await config()
}
