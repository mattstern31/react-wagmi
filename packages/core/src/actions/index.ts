export {
  connect,
  disconnect,
  fetchBalance,
  getAccount,
  getNetwork,
  signMessage,
  signTypedData,
  switchNetwork,
  watchAccount,
  watchNetwork,
  type ConnectArgs,
  type ConnectResult,
  type FetchBalanceArgs,
  type FetchBalanceResult,
  type GetAccountResult,
  type GetNetworkResult,
  type SignMessageArgs,
  type SignMessageResult,
  type SignTypedDataArgs,
  type SignTypedDataResult,
  type SwitchNetworkArgs,
  type SwitchNetworkResult,
  type WatchAccountCallback,
  type WatchNetworkCallback,
} from './accounts'

export {
  fetchToken,
  getContract,
  multicall,
  prepareWriteContract,
  readContract,
  readContracts,
  watchContractEvent,
  watchMulticall,
  watchReadContract,
  watchReadContracts,
  writeContract,
  type FetchTokenArgs,
  type FetchTokenResult,
  type GetContractArgs,
  type GetContractResult,
  type MulticallConfig,
  type MulticallResult,
  type PrepareWriteContractConfig,
  type PrepareWriteContractResult,
  type ReadContractConfig,
  type ReadContractResult,
  type ReadContractsConfig,
  type ReadContractsResult,
  type WatchContractEventConfig,
  type WatchContractEventCallback,
  type WatchMulticallConfig,
  type WatchMulticallCallback,
  type WatchReadContractConfig,
  type WatchReadContractCallback,
  type WatchReadContractsConfig,
  type WatchReadContractsCallback,
  type WriteContractArgs,
  type WriteContractMode,
  type WriteContractPreparedArgs,
  type WriteContractResult,
  type WriteContractUnpreparedArgs,
} from './contracts'

export {
  fetchEnsAddress,
  fetchEnsAvatar,
  fetchEnsName,
  fetchEnsResolver,
  type FetchEnsAddressArgs,
  type FetchEnsAddressResult,
  type FetchEnsAvatarArgs,
  type FetchEnsAvatarResult,
  type FetchEnsNameArgs,
  type FetchEnsNameResult,
  type FetchEnsResolverArgs,
  type FetchEnsResolverResult,
} from './ens'

export {
  fetchBlockNumber,
  fetchFeeData,
  watchBlockNumber,
  type FetchBlockNumberArgs,
  type FetchBlockNumberResult,
  type FetchFeeDataArgs,
  type FetchFeeDataResult,
  type WatchBlockNumberArgs,
  type WatchBlockNumberCallback,
} from './network-status'

export {
  getPublicClient,
  getWalletClient,
  getWebSocketPublicClient,
  watchPublicClient,
  watchWalletClient,
  watchWebSocketPublicClient,
  type GetPublicClientArgs,
  type GetPublicClientResult,
  type GetWalletClientArgs,
  type GetWalletClientResult,
  type GetWebSocketPublicClientArgs,
  type GetWebSocketPublicClientResult,
  type WatchWalletClientArgs,
  type WatchWalletClientCallback,
  type WatchPublicClientCallback,
  type WatchWebSocketPublicClientCallback,
} from './viem'

export {
  fetchTransaction,
  prepareSendTransaction,
  sendTransaction,
  waitForTransaction,
  watchPendingTransactions,
  type FetchTransactionArgs,
  type FetchTransactionResult,
  type PrepareSendTransactionArgs,
  type PrepareSendTransactionResult,
  type SendTransactionArgs,
  type SendTransactionResult,
  type WaitForTransactionArgs,
  type WaitForTransactionResult,
  type WatchPendingTransactionsArgs,
  type WatchPendingTransactionsCallback,
  type WatchPendingTransactionsResult,
} from './transactions'
