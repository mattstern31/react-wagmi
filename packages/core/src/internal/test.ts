export { getWalletClients, testChains } from '../../test'
