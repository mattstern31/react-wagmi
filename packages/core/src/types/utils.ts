/**
 * Converts properties of {@link T} to never
 *
 * @param T - Object to convert
 * @returns Object with properties converted to `never`
 *
 * @example
 * type Result = Never<{ foo: string, bar: number }>
 * //   ^? { foo: never, bar: never }
 */
export type Never<T> = {
  [K in keyof T]: never
}

/**
 * Makes {@link TKeys} optional in {@link TType} while preserving type inference.
 */
// s/o trpc (https://github.com/trpc/trpc/blob/main/packages/server/src/types.ts#L6)
export type PartialBy<TType, TKeys extends keyof TType> = Partial<
  Pick<TType, TKeys>
> &
  Omit<TType, TKeys>
