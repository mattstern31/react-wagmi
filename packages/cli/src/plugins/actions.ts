import { pascalCase } from 'change-case'
import dedent from 'dedent'

import type { Plugin } from '../config'
import type { RequiredBy } from '../types'
import { getAddressDocString, getIsPackageInstalled } from '../utils'

type ActionsConfig = {
  /**
   * Generate `getContract` action.
   *
   * @default true
   */
  getContract?: boolean
  /**
   * Override detected import source for actions.
   */
  overridePackageName?: 'wagmi/actions' | '@wagmi/core'
  /**
   * Generate `prepareWriteContract` action.
   *
   * @default true
   */
  prepareWriteContract?: boolean
  /**
   * Generate `readContract` action.
   *
   * @default true
   */
  readContract?: boolean
  /**
   * Generate `watchContractEvent` action.
   *
   * @default true
   */
  watchContractEvent?: boolean
  /**
   * Generate `writeContract` action.
   *
   * @default true
   */
  writeContract?: boolean
}

type ActionsResult = RequiredBy<Plugin, 'run'>

export function actions(config: ActionsConfig = {}): ActionsResult {
  const actions = {
    getContract: true,
    prepareWriteContract: true,
    readContract: true,
    watchContractEvent: true,
    writeContract: true,
    ...config,
  }
  return {
    name: 'Core',
    async run({ contracts, isTypeScript, outputs }) {
      const imports = new Set<string>([])
      const hasWriteContractMode = outputs.some(
        (x) =>
          x.plugin.name === 'React' && x.imports?.includes('WriteContractMode'),
      )

      const actionNames = new Set<string>()
      const getActionNameError = (name: string, contractName: string) =>
        new Error(
          `Action name "${name}" must be unique for contract "${contractName}".`,
        )

      const content: string[] = []
      for (const contract of contracts) {
        const baseActionName = pascalCase(contract.name)

        let typeParams = ''
        let omitted = ''
        const innerActionParams: Record<string, string> = {
          abi: contract.meta.abiName,
        }
        if (contract.meta.addressName) {
          omitted = `| 'address'`
          if (typeof contract.address === 'object') {
            typeParams = `& { chainId?: keyof typeof ${contract.meta.addressName} }`
            if (Object.keys(contract.address).length > 1) {
              innerActionParams.address = `${contract.meta.addressName}[config.chainId as keyof typeof ${contract.meta.addressName}]`
            } else
              innerActionParams.address = `${contract.meta.addressName}[${
                Object.keys(contract.address!)[0]
              }]`
          } else if (contract.address)
            innerActionParams.address = contract.meta.addressName
        }

        const innerActionConfig = `${Object.entries(innerActionParams).reduce(
          (prev, curr) => `${prev}${curr[0]}: ${curr[1]},`,
          '{',
        )}...config}`

        type Item = { name: string; value: string }
        const genDocString = (actionName: string, item?: Item) => {
          let description = `Wraps __{@link ${actionName}}__ with \`abi\` set to __{@link ${contract.meta.abiName}}__`
          if (item)
            description += ` and \`${item.name}\` set to \`"${item.value}"\``
          if (contract.address) {
            const docString = getAddressDocString({ address: contract.address })
            if (docString)
              return dedent`
              /**
              * ${description}.
              * 
              ${docString}
              */
              `
          }
          return dedent`
          /**
           * ${description}.
           */
          `
        }

        if (actions.getContract) {
          const name = `get${baseActionName}`
          if (actionNames.has(name))
            throw getActionNameError(name, contract.name)
          actionNames.add(name)

          imports.add('getContract')
          const docString = genDocString('getContract')

          let code
          if (isTypeScript) {
            imports.add('GetContractArgs')
            // prettier-ignore
            code = dedent`
            ${docString}
            export function ${name}(
              config: Omit<GetContractArgs, 'abi'${omitted}>${typeParams},
            ) {
              return getContract(${innerActionConfig})
            }
            `
          } else
            code = dedent`
            ${docString}
            export function ${name}(config) {
              return getContract(${innerActionConfig})
            }
            `
          content.push(code)
        }

        let hasReadFunction = false
        let hasWriteFunction = false
        let hasEvent = false
        for (const component of contract.abi) {
          if (component.type === 'function')
            if (
              component.stateMutability === 'view' ||
              component.stateMutability === 'pure'
            )
              hasReadFunction = true
            else hasWriteFunction = true
          else if (component.type === 'event') hasEvent = true
          // Exit early if all flags are `true`
          if (hasReadFunction && hasWriteFunction && hasEvent) break
        }

        if (hasReadFunction) {
          if (actions.readContract) {
            const name = `read${baseActionName}`
            if (actionNames.has(name))
              throw getActionNameError(name, contract.name)
            actionNames.add(name)

            imports.add('readContract')
            const docString = genDocString('readContract')

            let code
            if (isTypeScript) {
              imports.add('ReadContractConfig')
              code = dedent`
              ${docString}
              export function ${name}<
                TAbi extends readonly unknown[] = typeof ${contract.meta.abiName},
                TFunctionName extends string = string,
              >(
                config: Omit<ReadContractConfig<TAbi, TFunctionName>, 'abi'${omitted}>${typeParams},
              ) {
                return readContract(${innerActionConfig} as unknown as ReadContractConfig<TAbi, TFunctionName>)
              }
              `
            } else
              code = dedent`
              ${docString}
              export function ${name}(config) {
                return readContract(${innerActionConfig})
              }
              `
            content.push(code)
          }
        }

        if (hasWriteFunction) {
          if (actions.writeContract) {
            const name = `write${baseActionName}`
            if (actionNames.has(name))
              throw getActionNameError(name, contract.name)
            actionNames.add(name)

            imports.add('writeContract')
            const docString = genDocString('writeContract')

            let code
            if (isTypeScript) {
              const hasMultichainAddress = typeof contract.address === 'object'
              const TChainId = hasMultichainAddress
                ? `TMode extends WriteContractMode, TChainId extends number = keyof typeof ${contract.meta.addressName}`
                : ''
              let typeParams_ = ''
              if (TChainId) {
                if (!hasWriteContractMode) imports.add('WriteContractMode')
                typeParams_ = `& { mode: TMode; chainId?: TMode extends 'prepared' ? TChainId : keyof typeof ${contract.meta.addressName} }`
              }

              imports.add('WriteContractArgs')
              imports.add('WriteContractPreparedArgs')
              imports.add('WriteContractUnpreparedArgs')
              code = dedent`
              ${docString}
              export function ${name}<
                TFunctionName extends string,
                ${TChainId}
              >(
                config:
                  | (Omit<WriteContractPreparedArgs<typeof ${contract.meta.abiName}, TFunctionName>, 'abi'${omitted}>${typeParams_})
                  | (Omit<WriteContractUnpreparedArgs<typeof ${contract.meta.abiName}, TFunctionName>, 'abi'${omitted}>${typeParams_}),
              ) {
                return writeContract(${innerActionConfig} as unknown as WriteContractArgs<typeof ${contract.meta.abiName}, TFunctionName>)
              }
              `
            } else
              code = dedent`
              ${docString}
              export function ${name}(config) {
                return writeContract(${innerActionConfig})
              }
              `
            content.push(code)
          }

          if (actions.prepareWriteContract) {
            const name = `prepareWrite${baseActionName}`
            if (actionNames.has(name))
              throw getActionNameError(name, contract.name)
            actionNames.add(name)

            imports.add('prepareWriteContract')
            const docString = genDocString('prepareWriteContract')

            let code
            if (isTypeScript) {
              imports.add('PrepareWriteContractConfig')
              // prettier-ignore
              code = dedent`
              ${docString}
              export function ${name}<
                TAbi extends readonly unknown[] = typeof ${contract.meta.abiName},
                TFunctionName extends string = string,
              >(
                config: Omit<PrepareWriteContractConfig<TAbi, TFunctionName>, 'abi'${omitted}>${typeParams},
              ) {
                return prepareWriteContract(${innerActionConfig} as unknown as PrepareWriteContractConfig<TAbi, TFunctionName>)
              }
              `
            } else
              code = dedent`
              ${docString}
              export function ${name}(config) {
                return prepareWriteContract(${innerActionConfig})
              }
              `
            content.push(code)
          }
        }

        if (hasEvent) {
          if (actions.watchContractEvent) {
            const name = `watch${baseActionName}Event`
            if (actionNames.has(name))
              throw getActionNameError(name, contract.name)
            actionNames.add(name)

            imports.add('watchContractEvent')
            const docString = genDocString('watchContractEvent')

            let code
            if (isTypeScript) {
              imports.add('WatchContractEventConfig')
              imports.add('WatchContractEventCallback')
              // prettier-ignore
              code = dedent`
              ${docString}
              export function ${name}<
                TAbi extends readonly unknown[] = typeof ${contract.meta.abiName},
                TEventName extends string = string,
              >(
                config: Omit<WatchContractEventConfig<TAbi, TEventName>, 'abi'${omitted}>${typeParams},
                callback: WatchContractEventCallback<TAbi, TEventName>,
              ) {
                return watchContractEvent(${innerActionConfig} as WatchContractEventConfig<TAbi, TEventName>, callback)
              }
              `
            } else
              code = dedent`
              ${docString}
              export function ${name}(config, callback) {
                return watchContractEvent(${innerActionConfig}, callback)
              }
              `
            content.push(code)
          }
        }
      }

      let packageName
      if (config.overridePackageName) packageName = config.overridePackageName
      else if (await getIsPackageInstalled({ packageName: 'wagmi' }))
        packageName = 'wagmi/actions'
      else if (await getIsPackageInstalled({ packageName: '@wagmi/core' }))
        packageName = '@wagmi/core'
      else packageName = '@wagmi/core'

      const importValues = [...imports.values()]
      return {
        imports: importValues.length
          ? dedent`
            import { ${importValues.join(', ')} } from '${packageName}'
          `
          : '',
        content: content.join('\n\n'),
      }
    },
  }
}
