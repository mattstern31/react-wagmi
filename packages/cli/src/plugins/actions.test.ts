import { describe, expect, it } from 'vitest'

import { wagmiAbi } from '../../test'
import { format } from '../utils'
import { actions } from './actions'

describe('actions', () => {
  describe('run', () => {
    it('with TypeScript', async () => {
      const { imports, content } = await actions().run({
        contracts: [
          {
            name: 'Wagmi',
            address: '0xaf0326d92b97df1221759476b072abfd8084f9be',
            abi: wagmiAbi,
            content: '',
            meta: {
              abiName: 'WagmiAbi',
              addressName: 'WagmiAddress',
              configName: 'WagmiConfig',
            },
          },
        ],
        isTypeScript: true,
        outputs: [],
      })
      await expect(
        format(`${imports}\n\n${content}`),
      ).resolves.toMatchSnapshot()
    })

    it('without TypeScript', async () => {
      const { imports, content } = await actions().run({
        contracts: [
          {
            name: 'Wagmi',
            address: '0xaf0326d92b97df1221759476b072abfd8084f9be',
            abi: wagmiAbi,
            content: '',
            meta: {
              abiName: 'WagmiAbi',
              addressName: 'WagmiAddress',
              configName: 'WagmiConfig',
            },
          },
        ],
        isTypeScript: false,
        outputs: [],
      })
      await expect(
        format(`${imports}\n\n${content}`),
      ).resolves.toMatchSnapshot()
    })

    it('without TypeScript', async () => {
      const { imports, content } = await actions({
        overridePackageName: '@wagmi/core',
      }).run({
        contracts: [
          {
            name: 'Wagmi',
            address: '0xaf0326d92b97df1221759476b072abfd8084f9be',
            abi: [],
            content: '',
            meta: {
              abiName: 'WagmiAbi',
              addressName: 'WagmiAddress',
              configName: 'WagmiConfig',
            },
          },
        ],
        isTypeScript: false,
        outputs: [],
      })
      await expect(
        format(`${imports}\n\n${content}`),
      ).resolves.toMatchInlineSnapshot(`
        "import { getContract } from '@wagmi/core'

        /**
         * Wraps __{@link getContract}__ with \`abi\` set to __{@link WagmiAbi}__.
         */
        export function getWagmi(config) {
          return getContract({ abi: WagmiAbi, address: WagmiAddress, ...config })
        }
        "
      `)
    })

    it('throws for duplicate action names', async () => {
      await expect(
        actions().run({
          contracts: [
            {
              name: 'Wagmi',
              address: '0xaf0326d92b97df1221759476b072abfd8084f9be',
              abi: wagmiAbi,
              content: '',
              meta: {
                abiName: 'WagmiAbi',
                addressName: 'WagmiAddress',
                configName: 'WagmiConfig',
              },
            },
            {
              name: 'Wagmi',
              address: '0xaf0326d92b97df1221759476b072abfd8084f9be',
              abi: wagmiAbi,
              content: '',
              meta: {
                abiName: 'WagmiAbi',
                addressName: 'WagmiAddress',
                configName: 'WagmiConfig',
              },
            },
          ],
          isTypeScript: false,
          outputs: [],
        }),
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        '"Action name \\"getWagmi\\" must be unique for contract \\"Wagmi\\"."',
      )
    })
  })
})
