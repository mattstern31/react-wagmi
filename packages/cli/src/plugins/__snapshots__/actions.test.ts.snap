// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`actions > run > with TypeScript 1`] = `
"import {
  getContract,
  GetContractArgs,
  readContract,
  ReadContractConfig,
  writeContract,
  WriteContractArgs,
  WriteContractPreparedArgs,
  WriteContractUnpreparedArgs,
  prepareWriteContract,
  PrepareWriteContractConfig,
  watchContractEvent,
  WatchContractEventConfig,
  WatchContractEventCallback,
} from '@wagmi/core'

/**
 * Wraps __{@link getContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function getWagmi(config: Omit<GetContractArgs, 'abi' | 'address'>) {
  return getContract({ abi: WagmiAbi, address: WagmiAddress, ...config })
}

/**
 * Wraps __{@link readContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function readWagmi<
  TAbi extends readonly unknown[] = typeof WagmiAbi,
  TFunctionName extends string = string,
>(config: Omit<ReadContractConfig<TAbi, TFunctionName>, 'abi' | 'address'>) {
  return readContract({
    abi: WagmiAbi,
    address: WagmiAddress,
    ...config,
  } as unknown as ReadContractConfig<TAbi, TFunctionName>)
}

/**
 * Wraps __{@link writeContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function writeWagmi<TFunctionName extends string>(
  config:
    | Omit<
        WriteContractPreparedArgs<typeof WagmiAbi, TFunctionName>,
        'abi' | 'address'
      >
    | Omit<
        WriteContractUnpreparedArgs<typeof WagmiAbi, TFunctionName>,
        'abi' | 'address'
      >,
) {
  return writeContract({
    abi: WagmiAbi,
    address: WagmiAddress,
    ...config,
  } as unknown as WriteContractArgs<typeof WagmiAbi, TFunctionName>)
}

/**
 * Wraps __{@link prepareWriteContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function prepareWriteWagmi<
  TAbi extends readonly unknown[] = typeof WagmiAbi,
  TFunctionName extends string = string,
>(
  config: Omit<
    PrepareWriteContractConfig<TAbi, TFunctionName>,
    'abi' | 'address'
  >,
) {
  return prepareWriteContract({
    abi: WagmiAbi,
    address: WagmiAddress,
    ...config,
  } as unknown as PrepareWriteContractConfig<TAbi, TFunctionName>)
}

/**
 * Wraps __{@link watchContractEvent}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function watchWagmiEvent<
  TAbi extends readonly unknown[] = typeof WagmiAbi,
  TEventName extends string = string,
>(
  config: Omit<WatchContractEventConfig<TAbi, TEventName>, 'abi' | 'address'>,
  callback: WatchContractEventCallback<TAbi, TEventName>,
) {
  return watchContractEvent(
    {
      abi: WagmiAbi,
      address: WagmiAddress,
      ...config,
    } as WatchContractEventConfig<TAbi, TEventName>,
    callback,
  )
}
"
`;

exports[`actions > run > without TypeScript 1`] = `
"import {
  getContract,
  readContract,
  writeContract,
  prepareWriteContract,
  watchContractEvent,
} from '@wagmi/core'

/**
 * Wraps __{@link getContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function getWagmi(config) {
  return getContract({ abi: WagmiAbi, address: WagmiAddress, ...config })
}

/**
 * Wraps __{@link readContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function readWagmi(config) {
  return readContract({ abi: WagmiAbi, address: WagmiAddress, ...config })
}

/**
 * Wraps __{@link writeContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function writeWagmi(config) {
  return writeContract({ abi: WagmiAbi, address: WagmiAddress, ...config })
}

/**
 * Wraps __{@link prepareWriteContract}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function prepareWriteWagmi(config) {
  return prepareWriteContract({
    abi: WagmiAbi,
    address: WagmiAddress,
    ...config,
  })
}

/**
 * Wraps __{@link watchContractEvent}__ with \`abi\` set to __{@link WagmiAbi}__.
 */
export function watchWagmiEvent(config, callback) {
  return watchContractEvent(
    { abi: WagmiAbi, address: WagmiAddress, ...config },
    callback,
  )
}
"
`;
